mlflow/types/schema.py

```diff
@@ -406,3 +458,280 @@ def __eq__(self, other) -> bool:
+        :param value: parameter value
+        :param t: expected data type
+        """
+        if value is None:
+            return value
+
+        if not np.isscalar(value):
+            raise MlflowException(
+                f"Value should be a scalar for param {name}, got {value}", INVALID_PARAMETER_VALUE
+            )
+
+        if getattr(DataType, f"is_{t.name}")(value):
+            return value
+
+        if (
+            (DataType.is_integer(value) and t in (DataType.long, DataType.float, DataType.double))
+            or (DataType.is_long(value) and t in (DataType.float, DataType.double))
+            or (DataType.is_float(value) and t == DataType.double)
+            or t == DataType.datetime
+        ):
+            try:
+                if t == DataType.datetime:
+                    return np.datetime64(value)
+                return np.array(value, dtype=t.to_numpy()).take(0)
+            except ValueError as e:
+                raise MlflowException(
+                    f"Failed to convert value {value} from type {type(value).__name__} "
+                    f"to {t} for param {name}",
+                    INVALID_PARAMETER_VALUE,
+                ) from e
+
+        raise MlflowException(
+            f"Incompatible types for param {name}. Can not safely convert {type(value).__name__} "
+            f"to {t}.",
+            INVALID_PARAMETER_VALUE,
+        )
+
+    @classmethod
+    def validate_type_and_shape(
+        cls,
+        spec: str,
+        value: Union[DataType, List[DataType], None],
+        value_type: DataType,
+        shape: Optional[Tuple[int, ...]],
+    ):
+        """
+        Validate that the value has the expected type and shape.
+        """
+
+        def _is_1d_array(value):
+            if not isinstance(value, (list, np.ndarray)):
+                return False
+            if np.array(value).ndim == 1:
+                return True
+            return False
+
+        if shape is None:
+            return cls.enforce_param_datatype(f"{spec} with shape None", value, value_type)
+        elif shape == (-1,):
+            if not _is_1d_array(value):
+                raise MlflowException(
+                    f"Value must be a 1D array with shape (-1,) for param {spec}, "
+                    f"received {type(value).__name__} with ndim {np.array(value).ndim}",
+                    INVALID_PARAMETER_VALUE,
+                )
+            return [
+                cls.enforce_param_datatype(f"{spec} internal values", v, value_type) for v in value
+            ]
+        else:
+            raise MlflowException(
+                "Shape must be None for scalar value or (-1,) for 1D array value "
+                f"for ParamSpec {spec}), received {shape}",
+                INVALID_PARAMETER_VALUE,
+            )
+
+    @property
+    def name(self) -> str:
+        """The name of the parameter."""
+        return self._name
+
+    @property
+    def type(self) -> DataType:
```

Comment:
I think it would be a little less confusing to name this something other than the shadow of the builtin